# 01 Import Libraries

In [1]:
# import libraries
import pandas as pd
import numpy as np
import os

# 02 Import data

In [2]:
#import path of dataframe
path = r"C:\Users\bhimc\Documents\Instacart Basket Analysis (Career Foundry Rekha)"

In [3]:
#import "products.csv" from "Original Data"
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
#Import "orders_wrangles.csv" from "prepared data"
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 03 Consistancy Checking

## Adressing mixed-Typed Data

In [5]:
#create a dataframe
df_test = pd.DataFrame()

In [6]:
#Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [11]:
#Check for mixed types
#When running this code, it gave an error sayin that "dataFrame.applymap has been deprecated. Use dataFrame.map instead" so I changed the code and added ".map"
for col in df_test.columns.tolist():
    weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print(col)

mix


In [12]:
#Convert a column's data type from numeric to string
df_test['mix'] = df_test['mix'].astype('str')

## Missing Values

In [13]:
#Finding missing values in product.csv data frame
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [14]:
#Viewing the missing 16 values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [15]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [17]:
#Compare the number of rows in current dataframe with number of rows in the sebset
df_prods.shape

(49693, 5)

In [18]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [20]:
#Check that the number of rows has decreased
df_prods_clean.shape

(49677, 5)

## Duplicates

In [21]:
#Finding duplicated is dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [22]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [23]:
#Addressing Duplicates
df_prods_clean.shape

(49677, 5)

In [27]:
#Create new dataframe that doesn't address duplicates created previously
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [26]:
df_prods_clean_no_dups.shape

(49672, 5)

# 04 Task 4.5

## 2.Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function

In [28]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

## After analyzing my data, I do not see any discrepancy with the data

## 3."Check for mixed-type data in your df_ords dataframe."

In [31]:
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print(col)

## None of the columns have mix data since "col" wasn't printed

## 4. There aren't any mixed-data, so there is no need to fix it

## 5."Run a check for missing values in your df_ords dataframe."

In [35]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

## There are 206209 missing values in "days_since_prior order"

## 6. "Address the missing values using an appropriate method"

## Because there are a high number of missing values, it would not be good to delete them because it can reduce the data size and therefore impact it's statistics and lead to a wrong data analysis. A best course of action in this case would be to talk to the stakeholder.

In [42]:
#Instead of deleting the missing values in the dataframe, I decided to extract all the missing values to a new dataframe.
df_ords.shape

(3421083, 8)

In [39]:
df_ords_missing_values = df_ords[df_ords['days_since_prior_order'].isnull() == True]
# I chaged "isnull == false" to "isnull == true" because "true" would keep the missing values instead of deleting them.

## 7. "Run a check for duplicate values in your df_ords data."

In [40]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [41]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

## There are no duplicates found in the df_ords data frame.

## 8. "Address the duplicates using an appropriate method"

## Since there weren't any duplicates, no further course of action was necessary.

## 9. "Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names."

In [44]:
#Export df_prods_clean_no_dups as a new file
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_cleaned.csv'))

In [45]:
#Export df_ords as a new file
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_cleaned.csv'))